File: 课堂代码/曲线拟合/linearfit.m
%线性拟合
function [a,b] = linearfit (x, y)
    n = length (x);
    sx=sum(x);%所有自变量累加
    sy=sum(y);%所有因变量累加
    xx=x.*x;%所有自变量乘以自己集合
    xy=x.*y;%自变量和因变量一一对应相乘集合
    sxx=sum(xx);%自变量自乘之和
    sxy=sum(xy);%自变量乘以因变量之和
    deno=n*sxx-sx^2;%分母相同
    a=(n*sxy-sx*sy)/deno;%一次近似x前面的系数
    b=(sy*sxx-sxy*sx)/deno;
%     x2 = x.*x; 
%     xy = x.*y;
%     sx = sum(x); sy = sum(y);
%     sxy = sum (xy); sx2 = sum (x2);
%     deno = n*sx2-sx*sx;
%     a = (sy*sx2-sx*sxy)/deno;
%     b = (n*sxy-sx*sy)/deno;
end

File: 书中程序和图bmp figure code/chapter 3/chapter 3 code/3.2.2linear_fitting/linear_fit.cpp
/*linear fitting*/
#include <stdio.h>
main()
{
 int n =7,i;
 float x[]={0.5, 1.2, 2.1, 2.9, 3.6, 4.5, 5.7};
 float y[]={2.81,3.24,3.80,4.30,4.73,5.29,6.03};
 float sx=0.,sy=0.,sxy=0.,sx2=0,deno,a,b;
 
 for(i=0;i<=6;i++) 
 {
  sx=sx+x[i];
  sy=sy+y[i];
  sxy=sxy+x[i]*y[i];
  sx2=sx2+x[i]*x[i];
 }
 
 deno=n*sx2-sx*sx;
 a=(sy*sx2-sx*sxy)/deno;
 b=(n*sxy-sy*sx)/deno;

 printf("a=%6.2f    b=%6.2f\n",a,b);
 return(0);
}
